paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/FlinkSinkBuilder.java

```diff
@@ -282,4 +285,50 @@ private DataStream<RowData> trySortInput(DataStream<RowData> input) {
         }
         return input;
     }
+
+    private Integer checkAndUpdateParallelism(DataStream<?> input, Integer parallelism) {
+        try {
+            boolean parallelismUndefined = parallelism == null || parallelism == -1;
+            boolean isStreaming = FlinkSink.isStreaming(input);
+            boolean isAdaptiveParallelismEnabled =
+                    AdaptiveParallelism.isEnabled(input.getExecutionEnvironment());
+            boolean writeMCacheEnabled = table.coreOptions().writeManifestCache().getBytes() > 0;
+            boolean hashDynamicMode = table.bucketMode() == BucketMode.HASH_DYNAMIC;
+            if (parallelismUndefined
+                    && !isStreaming
+                    && isAdaptiveParallelismEnabled
+                    && (writeMCacheEnabled || hashDynamicMode)) {
+                List<String> messages = new ArrayList<>();
+                if (writeMCacheEnabled) {
+                    messages.add("Write Manifest Cache");
+                }
+                if (hashDynamicMode) {
+                    messages.add("Dynamic Bucket Mode");
+                }
```

Comment:
Add delimiter between "Write Manifest Cache" and "Dynamic Bucket Mode" if writeMCacheEnabled and hashDynamicMode are all true.

Reply:
The delimiter would be added during the message list's toString() method. Currently it looks like this

> Paimon Sink with [Write Manifest Cache, Dynamic Bucket Mode] does not support Flink's Adaptive Parallelism mode. Configuring sink parallelism to `input parallelism` instead. You can also set Paimon `sink.parallelism` manually to override this configuration.

Reply:
OK